chain/actors/builtin/miner/diff_deadlines.go

```diff
@@ -0,0 +1,180 @@
+package miner
+
+import (
+	"errors"
+
+	"github.com/filecoin-project/go-bitfield"
+	"github.com/filecoin-project/go-state-types/exitcode"
+)
+
+type DeadlinesDiff []*DeadlineDiff
+
+func DiffDeadlines(pre, cur State) (*DeadlinesDiff, error) {
+	changed, err := pre.DeadlinesChanged(cur)
+	if err != nil {
+		return nil, err
+	}
+	if !changed {
+		return nil, nil
+	}
+
+	numDl, err := pre.NumDeadlines()
+	if err != nil {
+		return nil, err
+	}
+	dlDiff := make(DeadlinesDiff, numDl)
```

Comment:
do we care about cases where a new deadline has been added in `cur` that was not in `pre`, or vice-versa?
Is there a reason to make `dlDiff` an array of all deadlines (with nil entries for those that have not changed) rather than a map of index->diff? if we expect sparse deadline changes, that might be a more natural representation.

Reply:
> do we care about cases where a new deadline has been added in cur that was not in pre, or vice-versa?

My understanding (is somewhat incomplete here, but I _think_) since [Deadlines are a fixed size](https://github.com/filecoin-project/specs-actors/blob/master/actors/builtin/miner/deadline_state.go#L63) it will never be the case that there exists a deadline in `cur` and not `pre`, or vice-versa, meaning we will walk each deadline on calls to [`ForEachDeadline`](https://github.com/filecoin-project/specs-actors/blob/master/actors/builtin/miner/deadline_state.go#L84). I suppose there could be an edge case here if we update the miner actor with a larger deadlines array and this code is called at the upgrade boundary -- I will leave a comment ackonwledging this as I am unsure how to address it...

I could be the case (perhaps this is what was meant) that there exists an [empty deadline](https://github.com/filecoin-project/specs-actors/blob/master/actors/builtin/miner/deadline_state.go#L120-L129) (a deadline with no partitions) in `cur` or `pre`, but I _think_ this addressed by walking both `cur` and `pre` in the `DiffDeadline` method. I'm still unsure on [this](https://github.com/filecoin-project/lotus/pull/4178/files#diff-10ae3287fcea32f26d66de940cc4d9abR62) case, meaning I need to add some tests.

> Is there a reason to make dlDiff an array of all deadlines (with nil entries for those that have not changed) rather than a map of index->diff? if we expect sparse deadline changes, that might be a more natural representation.

I don't have a strong reason for it, and agree a map would be better choice -- will address.

Reply:
the default case is that fixed-size list (all non-populated entries) are the same 'empty deadline', e.g. https://decentralized.app/?state=WyJiYWZ5MmJ6YWNlZHpqejJqZW94aHRoY3lyNDRxd3lmc2kyeHBmcnpsM2dmZDJjNDZoN3lqcG9lNDJ6bnQzNiIsW1swLDAsMCwwLDAsMCwwLCIiLDAse30sWzAseyI2Ijp7fX1dLHt9LHt9LHt9LHt9LHt9LHt9LHt9LHt9XV0sMCwwXQ==#storageMinerActor.Head.Deadlines.Due